lib/WasmReader/WasmBinaryReader.cpp

```diff
@@ -60,157 +69,208 @@ WasmBinaryReader::WasmBinaryReader(PageAllocator * alloc, byte* source, size_t l
-            {
-                ThrowDecodingError(L"Signatures section missing before function table");
-            }
-            FunctionHeader();
-            m_moduleState.count++;
-            break;
-
-        case bSectFunBodies:
-            if (!m_visitedSections->Test(bSectFunSigs))
-            {
-                ThrowDecodingError(L"Function signatures section missing before function bodies");
-            }
-            if (m_moduleInfo->GetFunctionCount() != m_moduleState.size)
-            {
-                ThrowDecodingError(L"Function signatures and function bodies count mismatch");
-            }
-            FunctionBodyHeader();
-            m_moduleState.count++;
-            return wnFUNC;
-
-        case bSectIndirectFunctionTable:
-            if (!m_visitedSections->Test(bSectFunSigs))
-            {
-                ThrowDecodingError(L"Function signatures section missing before indirect function table");
-            }
+        FunctionBodyHeader();
+        m_moduleState.count++;
 
-            for (UINT32 i = 0; i < m_moduleState.size; i++)
-            {
-                m_moduleInfo->AddIndirectFunctionIndex(ReadConst<UINT16>());
-            }
-            m_moduleState.count += m_moduleState.size;
-            break;
+    case bSectIndirectFunctionTable:
+        if (!m_visitedSections->Test(bSectFunctionSignatures))
+        {
+            ThrowDecodingError(L"Function signatures section missing before indirect function table");
+        }
 
-        case bSectEnd:
-            // One module per file. We will have to skip over func names and data segments.
-            m_pc = m_end; // skip to end, we are done.
-            return wnLIMIT;
+        for (UINT32 i = 0; i < m_moduleState.size; i++)
+        {
+            m_moduleInfo->AddIndirectFunctionIndex(ReadConst<UINT16>());
         }
+        m_moduleState.count += m_moduleState.size;
+        break;
+
+    case bSectEnd:
+        // One module per file. We will have to skip over func names and data segments.
+        m_pc = m_end; // skip to end, we are done.
+    default:
+        Assert(false);
+        return false;
     }
-     return wnLIMIT;
+
+    // Returns true if we read everything we need
+    return m_moduleState.secHeader.end == m_pc;
 }
 
-SectionCode
-WasmBinaryReader::SectionHeader()
+SectionHeader
+WasmBinaryReader::ReadSectionHeader()
 {
-    UINT len;
+    UINT len = 0;
     UINT32 sectionSize;
     UINT32 idSize;
 
+    SectionHeader header;
+    header.start = m_pc;
     sectionSize = LEB128(len);
+    header.end = m_pc + sectionSize;
+    if (header.end > m_end)
+    {
+        ThrowDecodingError(L"Section size is greater than the module size");
+    }
+
     idSize = LEB128(len);
+    if (sectionSize < idSize + len)
+    {
+        ThrowDecodingError(L"Invalid section size");
+    }
     const char *sectionName = (char*)(m_pc);
     m_pc += idSize;
+
     m_moduleState.count = 0;
     m_moduleState.byteLen = sectionSize;
```

Comment:
Can drop m_moduleState.byteLen in favor of header.start and .end since byteLen is now redundant.